slsa/provenance.go

```diff
@@ -21,28 +21,49 @@ import (
 
 	intoto "github.com/in-toto/in-toto-golang/in_toto"
 	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
-	"github.com/slsa-framework/slsa-github-generator/github"
 )
 
 const (
-	// GithubHostedActionsBuilderID is the builder ID for Github hosted actions.
+	// GithubHostedActionsBuilderID is a default builder ID for Github hosted actions.
 	GithubHostedActionsBuilderID = "https://github.com/Attestations/GitHubHostedActions@v1"
 )
 
 var githubComReplace = regexp.MustCompile(`^(https?://)?github\.com/?`)
 
-// HostedActionsProvenance generates an in-toto provenance statement in the SLSA
-// v0.2 format for a workflow run on a Github actions hosted runner.
-func HostedActionsProvenance(ctx context.Context, w WorkflowRun, c *github.OIDCClient) (*intoto.ProvenanceStatement, error) {
+// HostedActionsGenerator is a SLSA provenance generator for Github Hosted
+// Actions. Provenance is generated based on a "build type" which defines the
+// format for many of the fields in the provenance metadata. Builders for
+// different ecosystems (languages etc.) can implement a build type from
+// scratch or by extending GithubActionsBuild.
+type HostedActionsGenerator struct {
+	buildType BuildType
+	clients   ClientProvider
+}
+
+// NewHostedActionsGenerator returns a SLSA provenance generator for the given build type.
+func NewHostedActionsGenerator(bt BuildType) *HostedActionsGenerator {
+	return &HostedActionsGenerator{
+		buildType: bt,
+		clients:   &DefaultClientProvider{},
+	}
+}
+
+// Generate generates an in-toto provenance statement in SLSA v0.2 format.
+func (g *HostedActionsGenerator) Generate(ctx context.Context) (*intoto.ProvenanceStatement, error) {
```

Comment:
I'm thinking of maybe serializing to json and returning the `[]byte` here so it's more clear that you shouldn't be modifying the `ProvenanceStatement` after it's been generated. Fields should be modified by implementing your own build type.

Reply:
I think I'm going to keep this as is until I know a bit more about whether we're going to want to use the `attest` API from cosign and what that's going to look like.